regparser/notice/xml.py

```diff
@@ -41,6 +41,101 @@ def _set_date_attr(self, date_type, value):
+                    to reflect their parent/child relationships.
+        """
+
+        if not agencies:
+            # The FR Notice XML doesn't tend to have all the metadata we need
+            # contained within it, so don't try to parse that, just log an
+            # error.
+            logging.warn("Preprocessing notice: no agency metadata.")
+            return {}
+
+        # We need turn turn the references to parent_ids into a tree of dicts
+        # that contain subagencies in children fields:
+        for agency in agencies:
+            agency["children"] = []
+        agency_map = {agency["id"]: agency for agency in agencies}
+        child_keys = []
+        for key in agency_map:
+            agency = agency_map[key]
+            if agency.get("parent_id") and agency["parent_id"] in agency_map:
+                agency_map[agency["parent_id"]]["children"].append(agency)
+                child_keys.append(key)
+        for key in child_keys:
+            del agency_map[key]
+
+        def add_children(el, children):
+            """
+            Given an element and a list of children, recursively appends
+            children as EREGS_SUBAGENCY elements with the appropriate
+            attributes, and appends their children to them, etc.
+
+            :arg Element el:    The XML element to add child elements to.
+                                Should be either EREGS_AGENCY or
+                                EREGS_SUBAGENCY.
+            :arg list children: dict objects containing the agency information.
+                                Must have subagencies in `children` fields.
+
+            :rtype: XML Element
+            """
+            if children:
+                for agency in children:
+                    sub_el = etree.Element("EREGS_SUBAGENCY")
+                    sub_el.attrib["eregs-agency-name"] = str(agency["name"])
+                    sub_el.attrib["eregs-agency-raw-name"] = str(
+                        agency["raw_name"])
+                    sub_el.attrib["eregs-agency-id"] = str(agency["id"])
+                    if agency.get("children", []):
+                        sub_el = add_children(sub_el, agency["children"])
+                    el.append(sub_el)
+            return el
+
+        # Add the elements, starting with a parent ``EREGS_AGENCIES`` element.
+        agencies_el = etree.Element("EREGS_AGENCIES")
+        for agency_id in agency_map:
+            agency = agency_map[agency_id]
+            if not agency.get("parent_id"):
+                agency_el = etree.Element("EREGS_AGENCY")
+            else:
+                agency_el = etree.Element("EREGS_SUBAGENCY")
+            agency_el.attrib["eregs-agency-name"] = str(agency["name"])
+            agency_el.attrib["eregs-agency-raw-name"] = str(agency["raw_name"])
+            agency_el.attrib["eregs-agency-id"] = str(agency["id"])
+            add_children(agency_el, agency.get("children", []))
+            agencies_el.append(agency_el)
```

Comment:
This code quite similar to add_children. Any reason not to just make them all `EREGS_AGENCY` and start with `children=agencies`?